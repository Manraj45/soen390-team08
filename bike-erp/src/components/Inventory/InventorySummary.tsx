// DEPENDENCIES
import axios from "axios";
import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { Bar, Doughnut } from "@reactchartjs/react-chart.js";

// SERVICES
import { BACKEND_URL } from "../../core/utils/config";

// STYLES
import { Button, Card, CardActions, CardContent, Typography } from "@material-ui/core";
import useStyles from "./InventorySummaryStyles";

const InventorySummary = ({ account }: any) => {

  const classes = useStyles();
  const [allLocations, setAllLocations] = useState<any[]>([]);
  const [locationInventories, setLocationInventories] = useState<any[]>([]);
  const [compTypePerLoc, setCompTypePerLoc] = useState<any[]>([]);
  const [compCountPerLoc, setCompCountPerLoc] = useState<any[]>([]);
  const [countDataset, setCountDataset] = useState<any[]>([]);
  const reducer = (accumulator, currentValue) => accumulator + currentValue;

  // Fetch all locations
  useEffect(() => {
    axios.all([
      axios.get(`${BACKEND_URL}/components/locations/all`)
    ]).then((response) => {
      setAllLocations(response[0].data);
    });
  }, []);

  // Fetch inventory by location
  useEffect(() => {
    let promises : Promise<any>[] = [];
    setLocationInventories([]);
    for (let i = 0; i < allLocations.length; i++) {
      promises.push(axios
        .get(`${BACKEND_URL}/components/componentByLocation`, { params: {location: `${allLocations[i].location_name}` }})
        .then((response) => {
          setLocationInventories((inv) => [...inv, response.data]);
        })
      );
    }
    Promise.all(promises);
  }, [allLocations]);

  // Calculate inventory count
  useEffect(() => {
    let componentTypePerLocation : any[] = [];
    let componentCountPerLocation : any[] = [];

    function calculateInventoryCount() {
      for (let i = 0; i < locationInventories.length; i++) {

        const frames = locationInventories[i].filter(item => item.component_type === "FRAME");
        for (let i = 0; i < frames.length; i++) {
          frames[i] = frames[i].quantity;
        }

        const saddles = locationInventories[i].filter(item => item.component_type === "SEAT");
        for (let i = 0; i < saddles.length; i++) {
          saddles[i] = saddles[i].quantity;
        }

        const wheels = locationInventories[i].filter(item => item.component_type === "WHEEL");
        for (let i = 0; i < wheels.length; i++) {
          wheels[i] = wheels[i].quantity;
        }

        const driveTrains = locationInventories[i].filter(item => item.component_type === "DRIVE_TRAIN");
        for (let i = 0; i < driveTrains.length; i++) {
          driveTrains[i] = driveTrains[i].quantity;
        }

        componentTypePerLocation.push([
          frames.reduce(reducer, 0),
          saddles.reduce(reducer, 0),
          wheels.reduce(reducer, 0),
          wheels.reduce(reducer, 0),
          driveTrains.reduce(reducer, 0)
        ]);

        componentCountPerLocation.push([
          frames.reduce(reducer, 0),
          saddles.reduce(reducer, 0),
          wheels.reduce(reducer, 0),
          wheels.reduce(reducer, 0),
          driveTrains.reduce(reducer, 0)
        ].reduce(reducer, 0));
      }
    }
    calculateInventoryCount();
    setCompTypePerLoc(componentTypePerLocation);
    setCompCountPerLoc(componentCountPerLocation);
  }, [locationInventories]);

  // Generate dataset
  const dataLabels = ["Frames", "Saddles", "Handles", "Wheels", "Drivetrains"];

  const backgroundColors = [
    "rgba(255, 99, 132, 0.2)",
    "rgba(54, 162, 235, 0.2)",
    "rgba(255, 206, 86, 0.2)",
    "rgba(75, 192, 192, 0.2)",
    "rgba(153, 102, 255, 0.2)",
    "rgba(255, 159, 64, 0.2)"
  ];
  const borderColors = [
    "rgba(255, 99, 132, 1)",
    "rgba(54, 162, 235, 1)",
    "rgba(255, 206, 86, 1)",
    "rgba(75, 192, 192, 1)",
    "rgba(153, 102, 255, 1)",
    "rgba(255, 159, 64, 1)"
  ];

  // Generate dataset for component type by location
  useEffect(() => {
    setCountDataset([]);
    for (let i = 0; i < allLocations.length; i++) {
      setCountDataset((data) => [...data, {
        label: allLocations[i].location_name.toUpperCase(),
        backgroundColor: backgroundColors[i%backgroundColors.length],
        borderColor: borderColors[i%backgroundColors.length],
        borderWidth: 1,
        data: compTypePerLoc[i]
      }]);
    }
    // eslint-disable-next-line
  }, [compTypePerLoc])

  const countData = {
    labels: dataLabels,
    datasets: countDataset
  }

  const countOptions = {
    scales: {
      yAxes: [
        {
          ticks: {
            beginAtZero: true,
          },
        },
      ],
    },
  }

  // Generate dataset for component count by location
  let locationNames : any[] = [];
  for (let i = 0; i < allLocations.length; i++) {
    locationNames[i] = allLocations[i].location_name.toUpperCase();
  }

  let locationBgColors: any[] = [];
  let locationBorderColors : any[] = [];
  for(let i = 0; i < allLocations.length; i++) {
    locationBgColors.push(backgroundColors[i%backgroundColors.length]);
    locationBorderColors.push(borderColors[i%backgroundColors.length]);
  }

  const locationData = {
    labels: locationNames,
    datasets: [{
      label: "Inventory per lcation",
      data: compCountPerLoc,
      backgroundColor: locationBgColors,
      borderColor: locationBorderColors,
      borderWidth: 1
    }]
  }

  return(
    <div className={classes.root}>
      <Card variant="outlined">
        <CardContent style={{ paddingBottom: 0 }}>
          <Typography style={{ textTransform: "capitalize" }} variant="h5">
            Inventory Summary
          </Typography>
        </CardContent>
        <CardContent>
          <div>
            <Doughnut type="doughnut" data={locationData}/>
            <Bar type="bar" data={countData} options={countOptions}/>
          </div>
        </CardContent>
        <CardActions className={classes.seeMore}>
          <Button href="/inventory" size="medium" style={{color: "white"}}>SEE FULL INVENTORY</Button>
        </CardActions>
      </Card>
    </div>
  );
};

const mapStateToProps = (state: any) => {
  return {
    account: state.account.account
  };
};

export default connect(mapStateToProps)(InventorySummary);