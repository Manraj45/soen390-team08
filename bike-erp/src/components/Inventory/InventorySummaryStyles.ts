import { makeStyles } from "@material-ui/core";

const useStyles = makeStyles((theme) => ({
  root: {
    width: 500,
    height: "100%",
    [theme.breakpoints.down("sm")]: {
      marginTop: 30,
      maxWidth: 350
    },
    marginBottom: 30,
  },
  headerCell: {
    paddingTop: 0
  },
  seeMore: {
    backgroundColor: "#f15e32",
  }
}));

export default useStyles;
