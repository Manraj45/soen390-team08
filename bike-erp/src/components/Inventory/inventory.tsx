import Table from "@material-ui/core/Table";
import TableBody from "@material-ui/core/TableBody";
import TableCell from "@material-ui/core/TableCell";
import TableHead from "@material-ui/core/TableHead";
import TableRow from "@material-ui/core/TableRow";
import Axios from "axios";
import React, { useEffect, useState } from "react";
import useStyles from "../Inventory/inventoryStyles";

const Inventory: React.FC = () => {
  const [inventoryTable, setInventoryTable] = useState<any[]>([]);

  useEffect(() => {
    Axios.get("http://localhost:3001/components/").then((response) => {
      setInventoryTable(response.data);
      console.log(response.data);
    });
  }, []);

  const classes = useStyles();

  return (
    <React.Fragment>
      <div id="inventoryPageTest" className={classes.background}>
        <br></br>
        <div className={classes.title}>Inventory</div>
        <br></br>
        <Table size="small" className={classes.tableStyle}>
          <TableHead className={classes.tableHead}>
            <TableRow className={classes.topRow}>
              <TableCell className={classes.emptyCell} />
              <TableCell>Type</TableCell>
              <TableCell>Price</TableCell>
              <TableCell>Quantity</TableCell>
              <TableCell>Status</TableCell>
              <TableCell>Size</TableCell>
              <TableCell>Component Type</TableCell>
              <TableCell>Location</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {inventoryTable.map((row) => (
              <TableRow key={row.component_id}>
                <TableCell className={classes.tableHead} />
                <TableCell className={classes.innerTable}>
                  {row.component_type}
                </TableCell>
                <TableCell className={classes.innerTable}>
                  {row.price}
                </TableCell>
                <TableCell className={classes.innerTable}>
                  {row.quantity}
                </TableCell>
                <TableCell className={classes.innerTable}>
                  {row.component_status}
                </TableCell>
                <TableCell className={classes.innerTable}>{row.size}</TableCell>
                <TableCell className={classes.innerTable}>
                  {row.specificComponentType}
                </TableCell>
                <TableCell className={classes.innerTable}>
                  {row.location_name}
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </div>
    </React.Fragment>
  );
};

export default Inventory;
