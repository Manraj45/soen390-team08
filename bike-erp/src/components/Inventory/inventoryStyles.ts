import { makeStyles } from "@material-ui/core";

const useStyles = makeStyles((theme) => ({
  title: {
    textAlign: "center",
    fontWeight: "bold",
    fontFamily: "Inter",
    fontSize: "64",
  },
  background: {
    background: "#DDDDDD",
  },
  tableHead: {
    background: "#BFBFBF",
  },
  tableStyle: {
    width: "60%",
    marginLeft: "auto",
    marginRight: "auto",
  },
  innerTable: {
    background: "#FFFFFF",
  },
  topRow: {
    fontWeight: "bold",
    fontFamily: "Inter",
  },
  emptyCell: {
    background: "#DDDDDD",
  },
}));

export default useStyles;
