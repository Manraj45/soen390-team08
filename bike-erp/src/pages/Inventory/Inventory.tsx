// DEPENDENCIES
import axios from "axios";
import React, { useEffect, useState } from "react";

// STYLING
import { Table, TableHead, TableBody, TableRow, TableCell } from "@material-ui/core"
import useStyles from "./InventoryStyles";

/*
  The inventory page.
  This shows what a logged-in user's inventory contains.
*/
const Inventory: React.FC = () => {
  
  const [inventoryTable, setInventoryTable] = useState<any[]>([]);

  useEffect(() => {
    axios.get("http://localhost:3001/components/").then((response) => {
      setInventoryTable(response.data);
    });
  }, []);

  const classes = useStyles();

  return (
    <React.Fragment>
      <div id="inventoryPageTest" className={classes.background}>
        <br></br>
        <div className={classes.title}>Inventory</div>
        <br></br>
        <Table size="small" className={classes.tableStyle}>
          <TableHead className={classes.tableHead}>
            <TableRow className={classes.topRow}>
              <TableCell className={classes.emptyCell} />
              <TableCell>Type</TableCell>
              <TableCell>Price</TableCell>
              <TableCell>Quantity</TableCell>
              <TableCell>Status</TableCell>
              <TableCell>Size</TableCell>
              <TableCell>Component Type</TableCell>
              <TableCell>Location</TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {inventoryTable.map((row) => (
              <TableRow key={row.component_id}>
                <TableCell className={classes.tableHead} />
                <TableCell className={classes.innerTable}>
                  {row.component_type}
                </TableCell>
                <TableCell className={classes.innerTable}>
                  {row.price}
                </TableCell>
                <TableCell className={classes.innerTable}>
                  {row.quantity}
                </TableCell>
                <TableCell className={classes.innerTable}>
                  {row.component_status}
                </TableCell>
                <TableCell className={classes.innerTable}>{row.size}</TableCell>
                <TableCell className={classes.innerTable}>
                  {row.specificComponentType}
                </TableCell>
                <TableCell className={classes.innerTable}>
                  {row.location_name}
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </div>
    </React.Fragment>
  );
};

export default Inventory;
