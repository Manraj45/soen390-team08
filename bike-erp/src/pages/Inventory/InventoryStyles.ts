import { makeStyles } from "@material-ui/core";

const useStyles = makeStyles((theme) => ({
  title: {
    alignItems: "center",
    textAlign: "center",
    fontWeight: "bold",
    fontSize: "200%",
  },
  background: {
    background: "#f0f0f0",
    paddingLeft: "10px",
    paddingRight: "10px",
  },
  tableHead: {
    backgroundColor: "#f15e32",
    padding: "7px",
  },
  tableStyle: {
    width: "100%",
    marginLeft: "auto",
    marginRight: "auto",
  },
  innerTable: {
    background: "#FFFFFF",
  },
  topRow: {
    fontWeight: "bold",
    color: "white",
    background: "#f15e32",
    minWidth:"150px",
  },
  place: {
    maxHeight: "50px",
    width: "70%",
    marginLeft: "auto",
    marginRight: "auto",
    overflow: "auto",
    [theme.breakpoints.down("sm")]: {
      marginTop: "50px",
      minHeight: "500px",
    },
    [theme.breakpoints.up("md")]: {
      marginTop: "10px",
      minHeight: "650px",
    },
  },
}));

export default useStyles;
