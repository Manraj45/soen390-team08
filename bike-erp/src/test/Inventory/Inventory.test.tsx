import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import * as React from "react";
import renderer from "react-test-renderer";
import Inventory from "../../pages/Inventory/inventory";

Enzyme.configure({ adapter: new Adapter() });

describe("Inventory", () => {
  let wrapper: any;

  beforeEach(() => {
    wrapper = shallow(<Inventory />);
  });

  it("inventory page should appear", () => {
    console.log(wrapper);
    const data = wrapper.find("#inventoryPageTest");
    console.log(wrapper);
    expect(data.length).toBe(1);
  });

  it("inventory render", () => {
    expect(wrapper.exists()).toBe(true);
  });

  it("matches the snapshot", () => {
    const tree = renderer.create(<Inventory />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
