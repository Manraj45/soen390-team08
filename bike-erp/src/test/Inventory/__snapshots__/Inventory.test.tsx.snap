// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inventory matches the snapshot 1`] = `
<div
  className="makeStyles-background-2"
  id="inventoryPageTest"
>
  <br />
  <div
    className="makeStyles-title-1"
  >
    Inventory
  </div>
  <br />
  <div
    className="MuiPaper-root makeStyles-tableBack-3 MuiPaper-elevation1 MuiPaper-rounded"
  />
</div>
`;
